src/GitVersion.Core/VersionCalculation/VersionSearchStrategies/BaseVersionV2.cs

```diff
@@ -0,0 +1,87 @@
+namespace GitVersion.VersionCalculation;
+
+internal class BaseVersionV2 : BaseVersion
+{
+    internal static BaseVersionV2 ShouldIncrementFalse(string source, ICommit? baseVersionSource, string? label, SemanticVersion? alternativeSemanticVersion = null) => new(source, false)
+    {
+        BaseVersionSource = baseVersionSource,
+        Increment = VersionField.None,
+        SemanticVersion = null,
+        Label = label,
+        AlternativeSemanticVersion = alternativeSemanticVersion
+    };
+
+    internal static BaseVersionV2 ShouldIncrementFalse(string source, ICommit? baseVersionSource, SemanticVersion semanticVersion) => new(source, false)
+    {
+        BaseVersionSource = baseVersionSource,
+        Increment = VersionField.None,
+        SemanticVersion = semanticVersion
+    };
+
+    internal static BaseVersionV2 ShouldIncrementTrue(string source, ICommit? baseVersionSource, VersionField increment, string? label, bool forceIncrement, SemanticVersion? alternativeSemanticVersion = null) => new(source, true)
+    {
+        BaseVersionSource = baseVersionSource,
+        Increment = increment,
+        Label = label,
+        ForceIncrement = forceIncrement,
+        AlternativeSemanticVersion = alternativeSemanticVersion
+    };
+
+    public BaseVersionV2(string source, bool shouldIncrement) : base(source, shouldIncrement)
+    {
+    }
+
+    public BaseVersionV2(string source, bool shouldIncrement, SemanticVersion semanticVersion, ICommit? baseVersionSource, string? branchNameOverride)
+        : base(source, shouldIncrement, semanticVersion, baseVersionSource, branchNameOverride)
+    {
+    }
+
+    public bool ForceIncrement { get; init; }
+
+    public VersionField Increment { get; init; }
+
+    public string? Label { get; init; }
+
+    internal SemanticVersion? AlternativeSemanticVersion { get; init; }
```

Comment:
can you explain what is the `AlternativeSemanticVersion` for?

Reply:
Yes of course. In case of pre-release tagging you are sometimes only interested in tags with e.g. beta label. If you now find one with 1.0.0-beta.1 you are calculating the number of commits from this position. But if you hav other labels in between e.g. 1.1.0-alpha.1 then you want to consider this as well. So instead of yielting 1.0.0-beta.1+10 you are yielting 1.1.0.beta.1+10. Alternative means you have a tag but it's matching with the current label on branch.

Reply:
Also we can have more then one tag per commit.